src/details/hook_rooting.h

```diff
@@ -0,0 +1,254 @@
+
+// ReturnTypeTraits is an auxiliary traits class that determines the
+// return type of a hook member function.
+//
+template<typename T>
+struct ReturnTypeTraits {};
+
+template<typename R, typename T, typename... HookArgs>
+struct ReturnTypeTraits<R(T::*)(HookArgs...)> {
+  typedef R Type;
+};
+
+// The _HOOK_TASK macro defines an auxiliary classes HookTask_... that invoke
+// a specific plugin hook method with a provided set of method arguments. The
+// HookTask_... classes are meant to be used by HookLoop's apply function
+// called in the static functions of kaleidoscope::Hooks to forward the call to
+// the hook methods of all registered plugins.
+//
+#define _HOOK_TASK(HOOK_METHOD)                                                \
+__NN__                                                                         \
+__NN__   struct HookTask_##HOOK_METHOD {                                       \
+__NL__                                                                         \
+__NN__      /* Use a traits class to determine the return value type of the    \
+__NN__         hook method.                                                    \
+__NN__       */                                                                \
+__NL__      typedef typename ReturnTypeTraits<                                 \
+__NL__                 decltype(&kaleidoscope::Plugin::HOOK_METHOD)            \
+__NL__              >::Type  ReturnType;                                       \
+__NL__                                                                         \
+__NL__      template<typename Plugin__, typename... Args__>                    \
+__NL__      static ReturnType invoke(Plugin__ &plugin,                         \
+__NL__                                    Args__&&... hook_args)               \
+__NL__      {                                                                  \
+__NN__         /* Assert that hook methods in derived plugins have an          \
+__NN__            call signature.                                              \
+__NN__          */                                                             \
+__NL__         _MATCH_HOOK_TYPE(HOOK_METHOD, Plugin__)                         \
+__NN__                                                                         \
+__NL__         return plugin.HOOK_METHOD(hook_args...);                        \
+__NL__      }                                                                  \
+__NL__   };
+
+// A predicate class that decides on hook-plugin-loop continuation.
+//
+struct ContinueIfHookReturnsTrue {
+  static bool eval(bool val) {
+    return val;
+  }
+};
+
+// This macro implements the static hook rooting functions of
+// class kaleidoscope::Hooks based on the registered plugins.
+// It must be invoked in global namespace as it adds namespace kaleidoscope
+// to the Hooks' function implementations.
+//
+#define _HOOKS_STATIC_METHODS_IMPLEMENTATION                                   \
```

Comment:
This is the implementation of the static functions of class `Hook`. The implementation depends on the registered plugins that are considered by the `HookLoop` class. That is the reason why it must be implemented within a macro. The macro is called at the end of `_KALEIDOSCOPE_INIT_PLUGINS`.